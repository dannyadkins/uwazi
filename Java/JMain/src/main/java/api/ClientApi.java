package uwazi.es.api;

import com.googlecode.jsonrpc4j.JsonRpcParam;
import com.googlecode.jsonrpc4j.JsonRpcService;
import com.googlecode.jsonrpc4j.JsonRpcError;
import com.googlecode.jsonrpc4j.JsonRpcErrors;

import java.util.HashMap;
import com.google.common.collect.TreeMultimap;
import java.util.List;

@JsonRpcService("/client-api")
public interface ClientApi {

    @JsonRpcErrors({
    @JsonRpcError(exception=Throwable.class,
                  code=-1,
                  message="GetTokenUp threw an exception.",
                  data="")})
    byte[] GenTokenUp(@JsonRpcParam(value = "password") String password,
                      @JsonRpcParam(value = "keywords") String[] keywords,
                      @JsonRpcParam(value = "metadata") HashMap<String, String> metadata)
        throws Exception;


    @JsonRpcErrors({
    @JsonRpcError(exception=Throwable.class,
                  code=-1,
                  message="GenQueryToken threw an exception.",
                  data="")})
    byte[][] GenQueryToken(@JsonRpcParam(value = "password") String password,
                           @JsonRpcParam(value = "keyword") String keyword)
        throws Exception;

    
    @JsonRpcErrors({
    @JsonRpcError(exception=Throwable.class,
                  code=-1,
                  message="GetQueryToken threw an exception.",
                  data="")})
    List<String> Resolve(@JsonRpcParam(value = "password") String password,
                         @JsonRpcParam(value = "queryBytes") byte[][] queryBytes)
        throws Exception;
}