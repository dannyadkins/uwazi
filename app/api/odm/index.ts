/** @format */

export * from './model';
export * from './models';
