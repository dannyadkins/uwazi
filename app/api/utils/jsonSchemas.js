import { templateTypes } from 'shared/templateTypes';

const fieldTypes = Object.values(templateTypes);

export const objectIdSchema = {
  oneOf: [
    { type: 'string' },
    { type: 'object' }
  ]
};

export const linkSchema = {
  type: 'object',
  required: ['label', 'url'],
  properties: {
    label: { type: 'string', minLength: 1 },
    url: { type: 'string', minLength: 1 }
  }
};

export const dateRangeSchema = {
  type: 'object',
  properties: {
    from: {
      oneOf: [{ type: 'number' }, { type: 'null' }]
    },
    to: {
      oneOf: [{ type: 'number' }, { type: 'null' }]
    },
  },
  additionalProperties: false
};

export const latLonSchema = {
  type: 'object',
  required: ['lon', 'lat'],
  properties: {
    label: { type: 'string' },
    lat: { type: 'number', minimum: -90, maximum: 90 },
    lon: { type: 'number', minimum: -180, maximum: 180 }
  }
};

export const geolocationSchema = {
  type: 'array',
  items: latLonSchema
};

export const tocSchema = {
  type: 'object',
  properties: {
    range: {
      type: 'object',
      properties: {
        start: { type: 'number' },
        end: { type: 'number' }
      }
    },
    label: { type: 'string' },
    indentation: { type: 'number' }
  }
};

export const propertySchema = {
  type: 'object',
  required: ['label', 'type'],
  requireContentForSelectFields: true,
  requireRelationTypeForRelationship: true,
  requireInheritPropertyForInheritingRelationship: true,
  properties: {
    id: { type: 'string' },
    label: { type: 'string', minLength: 1 },
    name: { type: 'string', minLength: 1 },
    isCommonProperty: { type: 'boolean' },
    type: { type: 'string', enum: fieldTypes },
    prioritySorting: { type: 'boolean' },
    content: { type: 'string', minLength: 1 },
    inherit: { type: 'boolean' },
    inheritProperty: { type: 'string', minLength: 1 },
    filter: { type: 'boolean' },
    noLabel: { type: 'boolean' },
    fullWidth: { type: 'boolean' },
    defaultfilter: { type: 'boolean' },
    required: { type: 'boolean' },
    sortable: { type: 'boolean' },
    showInCard: { type: 'boolean' },
    style: { type: 'string' },
    nestedProperties: {
      type: 'array',
      items: {
        type: 'string'
      }
    }
  }
};
