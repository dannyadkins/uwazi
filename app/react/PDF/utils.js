/** @format */

import PDFJS from './PDFJS';
import superagent from 'superagent';
import CryptoJS from 'crypto-js';

function decrypt(transitmessage, pass) {
  var keySize = 256;
  var ivSize = 128;
  var iterations = 100;
  var salt = CryptoJS.enc.Hex.parse(transitmessage.substr(0, 32));
  var iv = CryptoJS.enc.Hex.parse(transitmessage.substr(32, 32));
  var encrypted = transitmessage.substring(64);

  var key = CryptoJS.PBKDF2(pass, salt, {
    keySize: keySize / 32,
    iterations: iterations,
  });

  var decrypted = CryptoJS.AES.decrypt(encrypted, key, {
    iv: iv,
    padding: CryptoJS.pad.Pkcs7,
    mode: CryptoJS.mode.CBC,
  });
  return decrypted.toString(CryptoJS.enc.Latin1);
}

function convertDataURIToBinary(raw) {
  var rawLength = raw.length;
  var array = new Uint8Array(new ArrayBuffer(rawLength));

  for (var i = 0; i < rawLength; i++) {
    array[i] = raw.charCodeAt(i);
  }
  return array;
}

const PDFUtils = {
  extractPDFInfo: pdfFile =>
    new Promise(resolve => {
      PDFJS.getDocument(pdfFile).promise.then(pdf => {
        const pages = [];
        for (let pageNumber = 1; pageNumber <= pdf.numPages; pageNumber += 1) {
          pages.push(pdf.getPage(pageNumber).then(PDFUtils.extractPageInfo));
        }

        return Promise.all(pages).then(result => {
          const count = {};
          result.forEach((length, index) => {
            count[index + 1] = {
              chars: length,
            };
            if (count[index]) {
              count[index + 1].chars += count[index].chars;
            }
          });
          resolve(count);
        });
      });
    }),

  extractEncryptedPDFInfo: pdfFile =>
    new Promise(resolve => {
      console.log('Extracting encrypted PDF data');
      superagent
        .get(pdfFile)
        .set('Accept', 'application/pdf')
        .set('X-Requested-With', 'XMLHttpRequest')
        .on('response', response => {
          console.log('RESPONSE:');
          var encryptedPdf = response.text;
          var decryptedPdf = decrypt(encryptedPdf, 'secret');
          console.log(decryptedPdf);
          var pdfAsArray = convertDataURIToBinary(decryptedPdf);

          PDFJS.getDocument({ data: decryptedPdf })
            .promise.then(pdf => {
              const pages = [];
              for (let pageNumber = 1; pageNumber <= pdf.numPages; pageNumber += 1) {
                pages.push(pdf.getPage(pageNumber).then(PDFUtils.extractPageInfo));
              }

              return Promise.all(pages)
                .then(result => {
                  const count = {};
                  result.forEach((length, index) => {
                    count[index + 1] = {
                      chars: length,
                    };
                    if (count[index]) {
                      count[index + 1].chars += count[index].chars;
                    }
                  });
                  resolve(count);
                })
                .catch(e => {
                  console.log(e);
                });
            })
            .catch(e => {
              console.log(e);
            });
        })
        .end();
    }),

  extractPageInfo: page =>
    new Promise(resolve => {
      const textLayerDiv = document.createElement('div');
      textLayerDiv.className = 'textLayer';

      page.getTextContent({ normalizeWhitespace: true }).then(textContent => {
        const textLayer = PDFJS.renderTextLayer({
          textContent,
          container: textLayerDiv,
          viewport: page.getViewport({
            scale: 1,
          }),
        });

        textLayer.promise.then(() => {
          resolve(textLayerDiv.innerText.length);
        });
      });
    }),
};

export default PDFUtils;
