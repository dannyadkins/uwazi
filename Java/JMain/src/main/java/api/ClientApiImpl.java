package uwazi.es.api;


// RPC stuff
import com.googlecode.jsonrpc4j.spring.AutoJsonRpcServiceImpl;
import org.springframework.stereotype.Service;

import uwazi.es.api.ObjSerializer;

// Clusion stuff
import org.crypto.sse.*;


// Everything else
import java.util.HashMap;
import com.google.common.collect.TreeMultimap;
import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.Multimap;


@Service
@AutoJsonRpcServiceImpl
public class ClientApiImpl implements ClientApi {
    private byte[] GetSkFromPassword(String password) throws Exception {
        return RR2Lev.keyGen(256, password, "salt/salt", 100000);
    }

    public byte[] GenTokenUp(String password,
               String[] keywords,
               HashMap<String, String> metadata) throws Exception {

        byte[] sk = GetSkFromPassword(password);

        String docId = metadata.get("docId");
        
        // TODO: Make multimap of {keyword: metadata} instead of {keyword: docId}
        Multimap<String, String> multimap = ArrayListMultimap.create();
        for (int i = 0; i < keywords.length; i++) {
            multimap.put(keywords[i], docId);
        }

        TreeMultimap<String, byte[]> tokenUp = DynRH.tokenUpdate(sk, multimap);
        
        // Serialize `tokenUp` into byte[] to be passed into RPC.
        return ObjSerializer.ToBytes(tokenUp);
    }
}