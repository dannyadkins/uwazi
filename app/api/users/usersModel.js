import mongoose from 'mongoose';

import { instanceModel } from 'api/odm';

const userSchema = new mongoose.Schema({
  username: { type: String, unique: true, required: true },
  password: { type: String, select: false, required: true },
  email: { type: String, unique: true, required: true },
  role: { type: String, unique: false, required: true },
  failedLogins: { type: Number, required: false, select: false },
  accountLocked: { type: Boolean, select: false },
  accountUnlockCode: { type: String, select: false }
});

export default instanceModel('users', userSchema);
