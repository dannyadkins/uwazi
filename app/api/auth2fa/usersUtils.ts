/** @format */

import * as otplib from 'otplib';

import usersModel, { User } from 'api/users/usersModel';
import { createError } from 'api/utils';

const checkUserExists = (user: User) => {
  if (!user) {
    throw createError('User not found', 403);
  }
};

export const setSecret = async (user: User) => {
  const [dbUser] = await usersModel.get({ _id: user._id });

  checkUserExists(dbUser);

  const secret = otplib.authenticator.generateSecret();
  const otpauth = otplib.authenticator.keyuri(dbUser.username || '', 'Uwazi', secret);

  if (!dbUser.using2fa) {
    await usersModel.save({ _id: dbUser._id, secret });
  }

  return { secret, otpauth };
};

export const enable2fa = async (user: User, token: string) => {
  const [dbUser] = await usersModel.get({ _id: user._id }, '+secret');

  checkUserExists(dbUser);

  const isValid = otplib.authenticator.verify({ token, secret: dbUser.secret });

  if (isValid) {
    return usersModel.save({ _id: dbUser._id, using2fa: true });
  }

  throw createError('The token does not validate against the secret key!', 409);
};
