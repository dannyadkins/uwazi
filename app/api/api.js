import languageMiddleware from './utils/languageMiddleware';
import activitylogMiddleware from './activitylog/activitylogMiddleware';
import elasticConfig from './config/elasticIndexes';

export default (app, server) => {
  //set db to use
  elasticConfig.index = elasticConfig[app.get('env')];

  //common middlewares
  app.use(languageMiddleware);
  app.use(activitylogMiddleware);

  //module routes
  //require('./auth/routes.js')(app);

  require('./socketio/middleware.js').default(server, app);
  require('./auth2fa/routes').default(app);
  require('./relationships/routes.js').default(app);
  require('./activitylog/routes.js').default(app);
  require('./users/routes.js').default(app);
  require('./templates/routes.js').default(app);
  require('./search/routes.js').default(app);
  require('./semanticsearch/routes.js').default(app);
  require('./thesauris/routes.js').default(app);
  require('./relationtypes/routes.js').default(app);
  require('./documents/routes.js').default(app);
  require('./contact/routes.js').default(app);
  require('./entities/routes.js').default(app);
  require('./pages/routes.js').default(app);
  require('./upload/routes.js').default(app);
  require('./settings/routes.js').default(app);
  require('./i18n/routes.js').default(app);
  require('./attachments/routes.js').default(app);
  require('./sync/routes.js').default(app);
  require('./swagger/swaggerconfig.js').default(app);
};
