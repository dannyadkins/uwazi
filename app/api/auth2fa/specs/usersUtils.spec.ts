/** @format */

import * as otplib from 'otplib';
import db from 'api/utils/testing_db';
import usersModel from 'api/users/usersModel';
import { createError } from 'api/utils';

import * as usersUtils from '../usersUtils';
import fixtures, { userId, secretedUserId } from './fixtures';

describe('auth2fa userUtils', () => {
  beforeEach(async () => {
    await db.clearAllAndLoad(fixtures);
  });

  afterAll(async () => {
    await db.disconnect();
  });

  const verifyUserNotFound = async (method: string) => {
    let action;

    switch (method) {
      case 'verifyToken':
        action = usersUtils.verifyToken;
        break;
      case 'enable2fa':
        action = usersUtils.enable2fa;
        break;
      case 'reset2fa':
        action = usersUtils.reset2fa;
        break;
      case 'setSecret':
      default:
        action = usersUtils.verifyToken;
        break;
    }

    try {
      await action({ _id: db.id() }, 'any token');
      fail('should throw error');
    } catch (e) {
      expect(e).toEqual(createError('User not found', 403));
    }
  };

  const verifyTokenMock = (params: { token: string; secret: string }) => {
    if (params.token === 'correctToken' && params.secret === 'correctSecret') {
      return true;
    }
    return false;
  };

  describe('setSecret', () => {
    it("should save the secret string on a non-previously-2fa'd user", async () => {
      const { secret, otpauth } = await usersUtils.setSecret({ _id: userId });
      const [secretedUser] = await usersModel.get({ _id: userId }, '+secret');
      expect(secretedUser.secret).toBe(secret);
      expect(otpauth).toBe(
        otplib.authenticator.keyuri(secretedUser.username || '', 'Uwazi', secret)
      );
    });

    it("should not change a secret on a previously-2fa'd user", async () => {
      await usersModel.save({ _id: userId, using2fa: true, secret: 'OLDSECRET' });
      try {
        await usersUtils.setSecret({ _id: userId });
        fail('Should throw error');
      } catch (e) {
        const [secretedUser] = await usersModel.get({ _id: userId }, '+secret');
        expect(secretedUser.secret).toBe('OLDSECRET');
        expect(e).toEqual(createError('Unauthorized', 401));
      }
    });

    it('should throw if user not found', async () => {
      await verifyUserNotFound('setSecret');
    });
  });

  describe('verfiyToken', () => {
    beforeEach(() => {
      spyOn(otplib.authenticator, 'verify').and.callFake(verifyTokenMock);
    });

    it('should verify with valid token', async () => {
      const response = await usersUtils.verifyToken({ _id: secretedUserId }, 'correctToken');
      expect(response.validToken).toBe(true);
    });

    it('should fail with 401 when invalid token', async () => {
      try {
        await usersUtils.verifyToken({ _id: secretedUserId }, 'incorrectToken');
        fail('Should throw error');
      } catch (e) {
        expect(e.code).toBe(401);
        expect(e.message).toMatch(/two-factor authentication failed/i);
      }
    });

    it('should throw if user not found', async () => {
      await verifyUserNotFound('verifyToken');
    });
  });

  describe('enable2fa', () => {
    beforeEach(() => {
      spyOn(otplib.authenticator, 'verify').and.callFake(verifyTokenMock);
    });

    it('should set "using2fa" to true if token matches', async () => {
      const savedUser = await usersUtils.enable2fa({ _id: secretedUserId }, 'correctToken');
      const [enabledUser] = await usersModel.get({ _id: secretedUserId });
      expect(savedUser.using2fa).toBe(true);
      expect(enabledUser.using2fa).toBe(true);
    });

    it('should not set "using2fa" and throw if token does not match', async () => {
      try {
        await usersUtils.enable2fa({ _id: secretedUserId }, 'incorrectToken');
        fail('Should throw an error');
      } catch (e) {
        expect(e.code).toBe(409);
        expect(e.message).toMatch(/token does not validate/i);
      }
    });

    it('should throw if user not found', async () => {
      await verifyUserNotFound('enable2fa');
    });
  });

  describe('reset2fa', () => {
    it('should set "using2fa" to false, and delete secret for sent user', async () => {
      await usersModel.save({ _id: secretedUserId, using2fa: true });
      const { _id, ...response } = await usersUtils.reset2fa({ _id: secretedUserId }); // eslint-disable-line
      const [resetUser] = await usersModel.get({ _id: secretedUserId }, '+secret');
      expect(response).toMatchSnapshot();
      expect(resetUser.using2fa).toBe(false);
      expect(resetUser.secret).toBe(null);
    });

    it('should throw if user not found', async () => {
      await verifyUserNotFound('reset2fa');
    });

    // it('should not set "using2fa" and throw if token does not match', async () => {
    //   try {
    //     await usersUtils.enable2fa({ _id: secretedUserId }, 'incorrectToken');
    //     fail('Should throw an error');
    //   } catch (e) {
    //     expect(e.code).toBe(409);
    //     expect(e.message).toMatch(/token does not validate/i);
    //   }
    // });

    // it('should throw if user not found', async () => {
    //   try {
    //     await usersUtils.enable2fa({ _id: db.id() }, 'any token');
    //     fail('should throw error');
    //   } catch (e) {
    //     expect(e).toEqual(createError('User not found', 403));
    //   }
    // });
  });
});
