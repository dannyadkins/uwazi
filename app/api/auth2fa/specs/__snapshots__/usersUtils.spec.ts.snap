// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth2fa userUtils reset2fa should set "using2fa" to false, and delete secret for sent user 1`] = `
Object {
  "email": "another@email.com",
  "role": "admin",
  "username": "otheruser",
  "using2fa": false,
}
`;
